crates/codegen/parser/runtime/src/cursor.rs

```diff
@@ -359,62 +359,53 @@ impl Cursor {
         false
     }
 
-    pub fn find_matching<R, F: Fn(&Cursor) -> Option<R>>(&mut self, filter_map: F) -> Option<R> {
-        while !self.is_completed {
-            if let Some(result) = filter_map(self) {
-                return Some(result);
-            }
-            self.go_to_next();
-        }
-
-        None
-    }
-
+    /// Finds the first node matching the given predicate.
+    ///
     /// In contrast to `Iterator::find_*`, this does not consume the first item when found.
-    fn find_noconsume<F: Fn(&Node) -> Option<R>, R>(&mut self, predicate: F) -> Option<R> {
+    pub fn find_matching<R, F: Fn(&Node) -> Option<R>>(&mut self, predicate: F) -> Option<R> {
```

Comment:
It just occurred to me that it's actually a `find_map` rather than a simple `find`. Could we change the name to reflect that?

On a separate note, I'd rather not expose this publicly - if a user wants to use this function already, they can call the existing `<Cursor as Iterator>::find_map` function (albeit consuming). I only kept the `find_noconsume` as internal API and implemented our public `find_{cursor,rule}_*` on top of it to retain the behaviour of non-consuming the first element as the TS API relies on that; Ideally I'd stick on the Rust side to what's usually expected, i.e. `Iterator` behaviour.

Reply:
Oh, I didn't know about the `result @ Some(_)` syntax. Will use it instead. Thanks!

Reply:
In that case, since we want to be more explicit about the "non-consuming" part, should we even use "find"?
It is a bit confusing to me for the same reason you mentioned above.

How about we use `go_to_next_token() -> bool` and `go_to_next_token_with_kind() -> bool` instead?
This way, it is clear that it matches our other `go_to_*()` APIs, and it is supposed to not consume, and the caller can fetch the result via `.node()` like the other APIs.
They can always use the iterator `find_*()` APIs to find and consume in one shot.

Reply:
Yep, this makes a lot of sense! Let's stick with `go_to` for the non-consuming APIs; if we also want to expose the `find_matching`->`go_to(_node)_matching`, I'd also change the predicate to return bool, instead (i.e. be like non-consuming `find` rather than `find_map`).